### PR TITLE
#73 - test documentation

### LINKED ISSUE #73: Test documentation
I would like to add a simple test that would take all (?) public methods from `src` directory and check if they are documented in `docs` directory.

### PR DESCRIPTION
**Pull request by JakubKermes**
Added test to check if every public function from traits is described in documentation.
This should close #73.

### CODE REVIEW

**Comment by EwelinaSkrzypacz** [comment]
BTW. Conflict here

File: docs/elements/view.html [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+<code-snippet gherkin="Then view response contains:
+            | key   | value     |
+            | title | Welcome   |
+            | name  | UserName  |" php="
+    /**
+     * @Then view response contains:
+     */
+    public function viewResponseContains(TableNode $table): void
+"></code-snippet>
```
Comment (EwelinaSkrzypacz) [inline docs/elements/view.html:76]: Please add new line at the end of the file

**Review by EwelinaSkrzypacz — COMMENTED** [review]
1. I don't see "Code blocks" on the docs menu 

image

2. Sometimes some bas happens with text color.

For example here, the word "Given" is not blue

image


Here are the same codeblock, but one is colorful and the second one is green

image

File: tests/DocumentationTest.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+use PHPUnit\Framework\TestCase;
+
+class DocumentationTest extends TestCase
+{
+    public function testIfDocumentationContainsAllPublicFunctions(): void
+    {
+        $functions = $this->getPublicFunctions();
+
+        foreach ($functions as $class => $methods) {
+            $trait = explode("\\", $class );
```
Comment (krzysztofrewak) [inline tests/DocumentationTest.php:15]:
```suggestion
            $trait = explode("\\", $class);
```

File: tests/DocumentationTest.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            $fileName = Str::slug(end($trait)) . ".html";
+            $filePath = "docs/elements/$fileName";
+
+            if (!file_exists($filePath)) {
+                if (str_ends_with($filePath, "httprequest.html")) {
```
Comment (krzysztofrewak) [inline tests/DocumentationTest.php:20]: Why not to rename files to be unified?

File: tests/DocumentationTest.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            }
+
+            foreach ($methods as $method) {
+                $containsFunction = strpos($fileContent, $method) !== false;
+                $this->assertTrue($containsFunction, "Function $method is not documented in $filePath.");
+            }
+        }
+    }
+
+    public function getClasses()
```
Comment (krzysztofrewak) [inline tests/DocumentationTest.php:48]:
```suggestion
    protected function getClasses(): void
```

File: tests/DocumentationTest.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        $classes = include "vendor/composer/autoload_classmap.php";
+        $bltTraits = [];
+
+        foreach (array_keys($classes) as $class) {
+            if (str_contains($class, "Blumilk\BLT\Features\Traits")) {
+                $bltTraits[] = $class;
+            }
+        }
+
+        return $bltTraits;
```
Comment (krzysztofrewak) [inline tests/DocumentationTest.php:59]: Oh no. Method is named `getClasses()` and returns `$bltTraits`? There's something wrong with naming here.

File: tests/DocumentationTest.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    public function getPublicFunctions()
```
Comment (krzysztofrewak) [inline tests/DocumentationTest.php:62]: Typing and encapsulation.

File: tests/DocumentationTest.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            }
+
+            $includedTraitMethodNames = [];
+
+            foreach ($includedTraitPublicMethods as $includedTraitMethod) {
+                $includedTraitMethodNames[] = $includedTraitMethod->getName();
+            }
+
+            foreach ($publicMethods as $publicMethod) {
+                if (!in_array($publicMethod->getName(), $includedTraitMethodNames, true)) {
```
Comment (krzysztofrewak) [inline tests/DocumentationTest.php:87]:
```suggestion
                if (!in_array($publicMethod->getName(), $includedTraitMethodNames, strict: true)) {
```

Please use named parameters for primitive values like `true`.

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: tests/DocumentationTest.php (hunk 3 above)
Comment (JakubKermes) [reply tests/DocumentationTest.php:20]: We prefer one Http file in documentation, but separating Into HttpRequest and HttpResponse might be easier to find what user is looking for. Should we separate it?

Comment (krzysztofrewak) [reply tests/DocumentationTest.php]: Then maybe some structure instead of four ifs would be better?

File: tests/DocumentationTest.php [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            }
+
+            $fileContent = file_get_contents($filePath);
+
+            if ($fileContent === false) {
+                $this->fail("Unable to read the file $filePath.");
+            }
+
+            foreach ($methods as $method) {
+                $containsFunction = strpos($fileContent, $method) !== false;
```
Comment (krzysztofrewak) [inline tests/DocumentationTest.php:38]: https://php.watch/versions/8.0/str_contains

File: tests/DocumentationTest.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        }
+    }
+
+    protected function getBltTraits(): array
+    {
+        $classes = include "vendor/composer/autoload_classmap.php";
+        $bltTraits = [];
+
+        foreach (array_keys($classes) as $class) {
+            if (str_contains($class, "Blumilk\BLT\Features\Traits")) {
```
Comment (krzysztofrewak) [inline tests/DocumentationTest.php:50]: Maybe https://www.php.net/manual/en/function.str-starts-with.php would be better?

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: tests/DocumentationTest.php (hunk 9 above)
Comment (krzysztofrewak) [reply tests/DocumentationTest.php:53]: Are you going to change that?

Comment (EwelinaSkrzypacz) [reply tests/DocumentationTest.php:53]: @JakubKermes

Comment (JakubKermes) [reply tests/DocumentationTest.php]: Done

File: docs/elements/code-blocks.html [hunk 10]
```diff
@@ -0,0 +1,49 @@
+<article-header>Code Blocks</article-header>
```
Comment (krzysztofrewak) [inline docs/elements/code-blocks.html:1]: File and header should be renamed.

File: tests/DocumentationTest.php [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+                $bltTraits[] = $class;
+            }
+        }
+
+        return $bltTraits;
+    }
+
+    protected function getPublicFunctions(): array
+    {
+        $traitClasses = $this->getBltTraits();
```
Comment (krzysztofrewak) [inline tests/DocumentationTest.php:63]: This is minor thing, but if `getPublicFunctions()` calls `getBltTraits()`, then the first one should be above second one.

File: tests/DocumentationTest.php [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+declare(strict_types=1);
+
+use Illuminate\Support\Str;
+use PHPUnit\Framework\TestCase;
+
+class DocumentationTest extends TestCase
+{
+    public function testIfDocumentationContainsAllPublicFunctions(): void
+    {
+        $functions = $this->getPublicFunctions();
```
Comment (krzysztofrewak) [inline tests/DocumentationTest.php:12]: What about hooks?

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

**Review by krzysztofrewak — APPROVED** [review]

**Review by EwelinaSkrzypacz — APPROVED** [review]

**Comment by EwelinaSkrzypacz** [comment]
Can we merge that @JakubKermes?

**Pull request merged**